x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalMathProcessor.java

```diff
@@ -24,13 +26,78 @@
     public enum BinaryOptionalMathOperation implements BiFunction<Number, Number, Number> {
 
         ROUND((l, r) -> {
-            double tenAtScale = Math.pow(10., r.longValue());
+            long rLong = r.longValue();
+            if (l instanceof Long || l instanceof Integer) {
+                long lLong = l.longValue();
+                if (lLong == 0L || rLong >= 0) {
+                    return lLong;
+                }
+
+                long digitsToRound = -rLong;
+                int digits = (int) (Math.log10(Math.abs(l.doubleValue())) + 1);
+                if (digits <= digitsToRound) {
+                    return 0L;
+                }
```

Comment:
Yes, exactly: `lLong == 0L` is a corner case for log10 indeed, but it's a single one and can be managed by itself (`round(0)` is always `0`).
After managing that case, the `log10` is not longer a problem, since the result R is always `1 < R < 19`, that are the possible number of digits of a Long. There cannot be overflows with Integer.MIN_VALUE then.

About the division loop, I'll do a microbenchmark, not sure it's faster ;-)

Reply:
Did some microbenchmarks (take it with a grain of salt)

For 1M Double roundings:
- `round(n, 0)` takes ~37ms with `Math.pow()` and ~30ms with a loop 
- `round(n, 3)` takes ~50ms with `Math.pow()` and ~35ms with a loop 
- `round(n, 10)` both take ~2sec 

So there is some advantage, but it's a matter of milliseconds out of millions of evaluations, that is probably negligible compared to the whole query execution time.

The implementation of Java Math.pow() is pretty complex and takes into account many corner cases, I guess it takes accuracy in strong consideration. I suspect a loop of divisions (or a loop to calculate the power) could lose some precision on very big values (not for Long, where we are dealing with pretty small numbers).

After all, IMHO we could keep relying on current implementation.

Reply:
After some more thinking, I suppose the best approach here is a brute force IF based approach for the really common cases (IMHO round zero to five decimal digits, but let's discuss it), and fall back to Math.pow() for the other cases.
This seems to give a performance boost up to 40% on the common cases (eg. the execution of `ROUND(num, 2)`) and practically no overhead on the other cases.
(again, take these numbers with a grain of salt, it's just microbenchmarks)

Reply:
I'm fine with having these optimizations in, they should address the most common cases. But just noting that my initial observation pertained to the `log10`, not `pow` function. (related, some [bit-twiddling fun](https://lemire.me/blog/2021/06/03/computing-the-number-of-digits-of-an-integer-even-faster/)). But also that's very optional.